src/main/java/algorithms/curated170/hard/WordSquaresTrieBacktracking.java

```diff
@@ -0,0 +1,93 @@
+package algorithms.curated170.hard;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class WordSquaresTrieBacktracking {
+   public static class Solution {
+       public final static int LETTER_a = 97;
+
+       public List<List<String>> wordSquares(String[] words) {
+           TrieNode root = buiildTrie(words);
+           int n = words[0].length();
+           TrieNode[] trieNodeState = new TrieNode[n];
+           Arrays.fill(trieNodeState, root);
+           List<List<String>> ans = new ArrayList<>();
+           dfs(0, new char[n][n], ans, trieNodeState);
+           return ans;
+       }
+       void dfs(int rowIndex, char[][] matrix, List<List<String>> ans, TrieNode[] trieNodeState) {
+           if (rowIndex == matrix.length) {
+               ans.add(convertMatrixToList(matrix, rowIndex));
+               return;
+           }
+           TrieNode[] trieNodeStateNext = new TrieNode[trieNodeState.length];
+           for (String candidate : trieNodeState[rowIndex].list) {
+
+               if (isSymmetric(matrix, candidate, rowIndex, trieNodeState, trieNodeStateNext)){
+                   dfs(rowIndex + 1, matrix, ans, trieNodeStateNext);
+               }
+           }
+       }
+
+       private boolean isSymmetric(char[][] matrix, String candidate, int rowIndex,
+                                   TrieNode[] trieNodeState, TrieNode[] trieNodeStateNext) {
+
+           int i = rowIndex;
+           while (i < matrix.length) {
+               char ch = candidate.charAt(i);
+               matrix[rowIndex][i] = ch;
+               if (i > rowIndex) {
+                   matrix[i][rowIndex] = ch;
+                   int letterIndex = letterIndex(ch);
+                   if (trieNodeState[i].children[letterIndex] == null) break;
+                   trieNodeStateNext[i] = trieNodeState[i].children[letterIndex];
```

Comment:
It seems that `trieNodeStateNext` doesn't do anything. If I'm not mistaken, we can remove it.

Reply:
It has a function as a kind of "backtracking" to the older "trieNodeState" keeping the older immutable.
But, you can try removing of course.

Reply:
I didn't see it being inputted into dfs. It makes more sense now, as we are traversing after some prefix already gone through.